mypyc/genops.py

```diff
@@ -1684,61 +1689,72 @@ def unbox_or_cast(self, src: Value, target_type: RType, line: int) -> Value:
-        while not symbol_found and index >= 1:
-            if symbol in self.environments[index].symtable:
-                symbol_found = True
-                rtype = self.environments[index].symtable[symbol].type
-            else:
-                index -= 1
+        env = self.add(GetAttr(base, ENV_ATTR_NAME, line))
+        assert isinstance(env.type, RInstance), '{} must be of type RInstance'.format(env)
 
-        if symbol_found:
-            # TODO: The below code will regenerate temporary references for the the same variables,
-            #       alone with the same environments for different variables. We should instead
-            #       only need to read the environment once per function.
+        for symbol, target in outer_env.symtable.items():
+            env.type.class_ir.attributes[symbol.name()] = target.type
+            symbol_target = AssignmentTargetAttr(env, symbol.name())
+            self.environment.add_target(symbol, symbol_target)
 
-            # Traverse through the environment classes and call GetAttr until the environment that
-            # contains the symbol is reached.
-            func_index = index - 1  # Adjust for no func_info for module top-level
-            self.func_infos[func_index].env_class.attributes[symbol.name()] = rtype
-            env = self.func_infos[-1].self_reg
-            for i in range(func_index, len(self.func_infos)):
-                env = self.add(GetAttr(env, ENV_ATTR_NAME, symbol.line))
-            return AssignmentTargetAttr(env, symbol.name())
+        return env
 
-        return None
+    def load_env_registers(self, fdef: FuncDef) -> None:
+        """Loads the registers for a given FuncDef.
+
+        Adds the arguments of the FuncDef to the environment. If the FuncDef is nested inside of
+        another function, then this also loads all of the outer environments of the FuncDef into
+        registers so that they can be used when accessing free variables.
+        """
+        self.add_args_to_env(fdef.arguments, fdef.line, local=True)
+
+        if self.fn_info.is_nested:
+            index = len(self.environments) - 2
 
-    def add_args_to_environment(self, args: List[Argument], line: int, local: bool = True) -> None:
+            # Load the first outer environment. This one is special because it gets saved in the
+            # FuncInfo instance's prev_env_reg field.
+            if index > 1:
+                outer_env = self.environments[index]
+                self.fn_info.prev_env_reg = self.load_outer_env(self.fn_info.self_reg,
+                                                                outer_env, fdef.line)
+                index -= 1
+
+            # Load the remaining outer environments into registers.
+            env_reg = self.fn_info.prev_env_reg
+            while index > 1:
+                outer_env = self.environments[index]
+                env_reg = self.load_outer_env(env_reg, outer_env, fdef.line)
+                index -= 1
```

Comment:
I think this could probably be restructured to have less duplication between the first outer env case and the rest, though I'm not totally sure it would be nicer.